src/modeling_library/dist/relabeled_distribution.jl

```diff
@@ -66,7 +66,7 @@ function logpdf(d::RelabeledDistribution{T, U}, x::T, base_args...) where {T, U}
                 push!(logprobs, logpdf(d.base, index, base_args...))
             end
         end
-        logsumexp(logprobs)
+        isempty(logprobs) ? -Inf : logsumexp(logprobs)
```

Comment:
@ztangent Are there cases where logprobs being empty is not an error?

Reply:
Good question! I think this hinges on whether we want to allow users to evaluate the logpdf values for `labeled_cat` or `labeled_unif` distributions which may not have constant support, or if we want an error to be thrown. For clarity, by those two distributions, I mean:
```
@dist labeled_cat(labels, probs) = labels[categorical(probs)]
@dist labeled_unif(labels) = labels[uniform_discrete(1, length(labels))]
```
I have at least one use case where allowing choices with non-constant support might be justified, and where having `logpdf` return `-Inf` for those choices instead of erroring would be convenient for importance sampling. I'll put this in a separate comment below.

Reply:
Also, I think if we make `logpdf` throw errors for `labeled_cat` and `labeled_unif` when we query values outside their support, then I think we should make `categorical` and `uniform_discrete` do the same for consistency. We might also then want to do the same for all the other discrete distributions (`geometric` and `poisson`). But I wonder if this would be too restrictive...?

Reply:
Okay @alex-lew has a more concrete example of when the support of a `labeled_cat` distribution might be not constant, will let him add it below instead of mine! (Mine was a random graph search example.)

Reply:
My first thought on this is that it makes sense for `labeled_cat(labels, probs)`'s `logpdf` to throw an error when the input is not an element of `labels`, as @marcoct seems to also be suggesting above.

I think we need to be more precise about what "support" means.  See [Wikipedia](https://en.wikipedia.org/wiki/Support_(measure_theory)).  For `labeled_cat`, I would say that `labels` is the sample space, not necessarily the support.  The support is
```julia
[label for (label, prob) in zip(labels, probs) if prob > 0]
```
(Note that support is a bit more complicated in the continuous case, as there is a topological closure, which has no effect in the discrete case.)

I would say that `labeled_cat` defines a measure on the finite set `labels`, and the only things that measure should know how to talk about are subsets of `labels`.  Hence I think evaluating `logpdf(..., x)` where `x ∉ labels`, which is the same as evaluating `μ({x})` where `{x} ⊄ labels`, should give an error.

@ztangent, if I understand your above comments correctly, it sounds like you want an implicit conversion of measures via "extension by zero," where the Gen interpreter automatically converts μ to a measure ũ which has sample space `labels ∪ {x}`, with `ũ({x}) = 0` and agrees with μ on subsets of `labels` (these conditions uniquely determine ũ).  Perhaps what I'm trying to say is, I don't think that conversion should be implicit -- I suspect it would down the road suffer from the usual gotchas of generous implicit conversions, namely confusing behaviors and silent failures.

On the other hand, an ability to do _explicit_ conversion sounds totally fine, and may achieve the best of both worlds.  Concretely this would mean having a function in the GFI that implements
```
logpdf_extended(x) := (x in sample_space ? logpdf(x) : 0)
```

@ztangent, @marcoct, does the above make sense?

Reply:
@bzinberg Why not define the sample space of `labeled_cat` to be all Julia values (with the discrete sigma-algebra)? This involves no implicit conversions: it's just how it's defined.

The example that Xuan mentioned above is this:

Suppose I have a generative function that samples a state in the US, and then a city within that state:

```Julia
@dist labeled_cat(labels, frequencies) = labels[categorical(frequencies ./ sum(frequencies))]
@gen function city_and_state()
  state ~ labeled_cat(all_states, state_populations)
  city ~ labeled_cat(cities_within_state[state], city_populations_in_state[state])
end
```

I'd like to observe the `city` and infer the `state`.

Now, technically, the above generative function is "incorrect" in that it uses a single address, `city`, for choices that can have different supports. But if I did use a different address for `city` depending on the state, I wouldn't be able to _observe_ `city` without knowing the state.

Reply:
A different way of implementing `logpdf_extended`, which actually might be better, is to just implement a proper extension-by-zero of measures.  In this approach, the measure μ and a set A ⊇ sample_space(μ) are specified, and the output is a new measure,
```
ũ = extend_by_zero(μ, A)
```
such that ũ agrees with μ on subsets of `sample_space(μ)` and ũ(B) = 0 for all B ⊆ A.

I think this formulation does both what @marcoct wants and what @ztangent wants in a principled way:

* It still catches the kind of errors @marcoct wants it to catch
* It allows the user to bypass the check by explicitly declaring that `logpdf` of their given value should be computed via extension by zero if necessary.

Note also that it's possible to construct a categorical with big sample space and small support ahead of time, instead of extending it at runtime.  For example, if you want a discrete uniform distribution whose support is not known until runtime, you could have its sample space in every code path be all of ℕ, and have its `logpdf(x)` return 0 whenever `x > length(probs)`.

The analogous thing for `labeled_cat`, or any other discrete distribution, would be to have the set A above be specified by a predicate φ, that is, `φ(x) == true` iff x ∈ A.  Then the user gets to control how permissive the error checking is.  In @ztangent's case, φ could return true for everything, or could return true whenever `x` looks reasonably close to having the right format.  Making φ stricter results in more error checking, at the cost of having to write a stricter φ (which is trivial in many cases, esp. finite or integer-valued ones, but is nontrivial for some cases of infinite discrete sets of structured objects).

Reply:
@alex-lew, didn't see your response before writing my follow-on above, but I think the follow-on addresses your question :slightly_smiling_face:

Reply:
@bzinberg I think that's an interesting proposal. What about the converse, `restrict_sample_space`? This could be like an assertion, telling Gen that one would like to receive errors for values that fail a certain predicate.

Reply:
> @bzinberg I think that's an interesting proposal. What about the converse, `restrict_sample_space`? This could be like an assertion, telling Gen that one would like to receive errors for values that fail a certain predicate.

If we want `restrict_sample_space(μ)` to still be a probability measure, then this is the same thing as suggesting that we add the ability to condition on arbitrary predicates.  If we don't require that, i.e. we keep the values of `logpdf` just as they are, then I think it should literally, as you say, be written as an assertion, rather than building it into an object that has mathematical semantics of "measure that is not a probability measure" (so that we avoid including arbitrary measures into the conceptual machinery of Gen).

Reply:
> Mostly, I think we need to think carefully about what the default behavior should be.

Definitely -- and more strongly, I think it's important that we _document_ the default behavior, ideally also describing the mathematical formalism it corresponds to.

Reply:
> Why shouldn't the implicit sample space of `labeled_cat` be all objects, given that its argument could be changing from execution to execution?

I think you answered this question in the trace types paper?  We're trying to allow upper bounds on the data type of `logpdf` and friends.

Reply:
> If we want restrict_sample_space(μ) to still be a probability measure, then this is the same thing as suggesting that we add the ability to condition on arbitrary predicates.

Sure -- we could say that `restrict_sample_space(predicate)` actually restricts the sample space to the _union_ of the support and the predicate, which doesn't then require any conditioning. Or we could use a more accessible name like `add_support_assertion`. 

In general, I think anything that makes reasonable sense to us as an implementation strategy can be given a reasonable mathematical description. The most important bit to me is that we should think about what default behavior would be most useful for users.

Reply:
> And in practice, would 0-logpdf not serve as a red flag?

I don't think it should -- e.g. why should be considered bad practice for the user to declare a categorical random variable where some of the weights (perhaps computed from other runtime quantities, or via a neural net) are zero?

Reply:
> I think you answered this question in the trace types paper? We're trying to allow upper bounds on the argument data type of logpdf and friends.
 
In the trace types paper, we required that the support matched the sample space (we made no distinction).  For example, the trace types paper actually forbids a categorical variable with some of the weights being zero. Gen has typically forbidden it too, unless you make sure that the address of the random choice clearly indicates the support (in which case you should still never hit an `-Inf` score). 

The trace types paper is one especially restrictive point in the design space. Some of this was because the space restrictions in the paper didn't allow us to go into detail about designs that were less restrictive. I think this discussion is related to the trace types paper, but that Gen shouldn't necessarily take its cues from STPL.

Reply:
> What are realistic errors someone might make that we want to prevent against? Mis-spelled observations?

Misspelled is one case.  Another class of cases is bugs in the code that computes an `x` and evaluates `logpdf(x)`.  Similar to how we may want to hear complaints if we try to evaluate `[1, 2, 3] + 4`, we may want to hear complaints if we evaluate `logpdf(x)` where `x` is a string and the categorical distribution is defined over integers.

Reply:
@bzinberg it's worth noting that `logpdf` for `categorical` currently only accepts `Integer` values for `x`, so a type error will be thrown if you pass in a string!

Currently, user-defined distributions defined by the `@dist` DSL don't support this sort of type checking, but it looks like @alex-lew nicely implemented things so that it *should* be possible with a minor modification for type annotations to be supported when defining `@dist` distributions, e.g.:

```julia
@dist typed_labeled_cat(labels::String, probs) = labels[categorical(probs)]
```

I'll create an issue to note that typing is not possible right now.